File: Java/week1/FruitInventory.java
/*
 Filename: FruitInventory.java
 Author: Mikhail Zubko
 Created: 2025/01/16
 Purpose: 
*/

import java.util.HashMap;
import java.util.Scanner;

public class FruitInventory {
  protected static int getInput(Scanner keyboard, String prompt) {
    int result;
    while (true) {
      try {
        System.out.print(prompt);
        result = keyboard.nextInt();
        if (result < 0) {
          System.out.println("Enter a positive number.");
          continue;
        }
        return result;
      } catch (Exception e) {
        System.out.println("Enter a whole number.");
        keyboard.nextLine();
      }
    }
  }

  static void showMap(HashMap<String, Integer> map) {
    for (String entry : map.keySet()) {
      // use get(entry) to retrive the value
      System.out.println("Key: " + entry + ", Value: " + map.get(entry));
    }
  }

  public static void main(String[] args) {
    Scanner keyboard = new Scanner(System.in);
    HashMap<String, Integer> myMap = new HashMap<String, Integer>();
    char choice;
    String fruit;
    int quantity;

    System.out.println("--- Welcome to my Fruit Inventory System! ---");

    while (true) {
      System.out.print("\nEnter a fruit name to add: ");
      fruit = keyboard.nextLine();
      quantity = getInput(keyboard, "Enter the quantity for " + fruit + ": ");
      myMap.put(fruit, quantity);
      System.out.println("\nUpdated Fruit Inventory: ");
      showMap(myMap);
      System.out.print("\nDo you want to add more fruits? (y/n): ");
      choice = Character.toLowerCase(keyboard.next().charAt(0));
      if (choice != 'y') {
        System.out.println("\nFinal Fruit Inventory: ");
        showMap(myMap);
        break;
      }
      keyboard.nextLine();
    }
  }
}

File: Cpp/week2/fruit_inventory/fruit_inventory.cpp
#include <iostream>
#include <unordered_map>

void showMap(std::unordered_map<std::string, int> umap);

int main() {
  std::unordered_map<std::string, int> inventory;
  std::string fruit;
  std::string dummy; // to consume newlines
  int quantity;
  char choice;
  std::cout << "--- Welcome to my Fruit Inventory System! ---" << std::endl;
  while (true) {
    quantity = 0;
    std::cout << "Enter a fruit name to add: ";
    std::getline(std::cin, fruit);
    std::cout << "Enter the quantity for " << fruit << ": ";
    std::cin >> quantity;
    if (quantity < 1) {
      std::cout << "Wrong input." << std::endl;
      std::cin.clear();              // clear the error flag
      std::getline(std::cin, dummy); // clear the input
      continue;
    }
    inventory[fruit] = quantity;
    std::cout << std::endl;
    std::cout << "Updated Fruit Inventory:" << std::endl;
    showMap(inventory);
    std::cout << std::endl;
    std::cout << "Do you want to add more fruits? (y/n): ";
    std::cin >> choice;
    std::getline(std::cin, dummy); // clear the input
    std::cout << std::endl;
    if (choice != 'y') {
      std::cout << "Final Fruit Inventory:" << std::endl;
      showMap(inventory);
      break;
    }
  }

  return 0;
}
void showMap(std::unordered_map<std::string, int> umap) {
  for (auto x : umap) {
    std::cout << "Key: " << x.first << ", Value: " << x.second << std::endl;
  }
}
